File: DynamicProgramming/18largestIncreasingSubSeq/longestIncreasingSubsequence.java
public class longestIncreasingSubsequence {
    public static int lengthOfLIS(int[] nums) {
        int maxLen = 1;
        int[] lenLis = IntStream.generate(() -> 1).limit(nums.length).toArray();
        for (int i = nums.length; i > -1; i--) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] < nums[j]) {
                    lenLis[i] = lenLis[i] < 1 + lenLis[j] ? 1 + lenLis[j] : lenLis[i];
                    maxLen = maxLen > lenLis[i] ? maxLen : lenLis[i];
                }
            }
        }
        return maxLen;
    }

    public static void main(String[] args) {
        int[] nums = { 4, 10, 4, 3, 8, 9 };
        System.out.println(lengthOfLIS(nums));
    }
}

File: DynamicProgramming/18largestIncreasingSubSeq/longestIncreasingSubsequence.py
class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        maxLen = 1
        LenLis = [1] * len(nums)
        for i in range(len(nums) - 1, -1, -1):
            for j in range(i+1, len(nums)):
                if nums[i] < nums[j]:
                    LenLis[i] = max(LenLis[i], 1 + LenLis[j])
                    maxLen = max(maxLen, LenLis[i])                
        return maxLen
print(Solution().lengthOfLIS([10,9,2,5,3,7,101,18]))
print(Solution().lengthOfLIS([4,10,4,3,8,9]))
print(Solution().lengthOfLIS([0,1,0,3,2,3]))
print(Solution().lengthOfLIS([7,7,7,7,7,7,7]))